File: typescript/day1/part2/src/main.ts
async function main() {
  const text = await Bun.file("src/data.txt").text();
  const lines = text.split("\n");

  const calibrartion = lines.reduce((acc, line) => {
    let firstNum = null;
    let secondNum = null;

    for (var i = 0; i < line.length; i++) {
      const num = line.charAt(i);
      const numSubString = line.substring(i);

      const numString = stringNumber(numSubString);

      if (numString != null) {
        if (firstNum == null) firstNum = numString;
        secondNum = numString;
        continue;
      }

      if (!isNaN(Number(num))) {
        if (firstNum == null) firstNum = num;
        secondNum = num;
      }
    }

    const combined = firstNum! + secondNum!;

    return acc + Number(combined);
  }, 0);

  console.log(calibrartion);
}

function stringNumber(input: string): string | null {
  const keys = Object.keys(numbers);
  for (var i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (input.startsWith(key)) {
      return numbers[key];
    }
  }

  return null;
}

const numbers: { [key: string]: string } = {
  one: "1",
  two: "2",
  three: "3",
  four: "4",
  five: "5",
  six: "6",
  seven: "7",
  eight: "8",
  nine: "9",
};

main();

File: rust/day1/part2/src/main.rs
fn main() {
    let contents = include_str!("data.txt");

    const NUMBERS: &[&str] = &[
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    let mut total = 0;
    for line in contents.lines() {
        let mut first_number = None;
        let mut last_number = None;

        for (i, c) in line.chars().enumerate() {
            if c.is_numeric() {
                match first_number {
                    Some(_) => {}
                    None => {
                        first_number = Some(c);
                    }
                }
                last_number = Some(c);
            } else {
                let sub_string = line[i..].to_string();

                for (i, number) in NUMBERS.iter().enumerate() {
                    if sub_string.starts_with(number) {
                        match first_number {
                            Some(_) => {}
                            None => {
                                first_number = Some(format!("{}", i + 1).parse::<char>().unwrap())
                            }
                        }
                        last_number = Some(format!("{}", i + 1).parse::<char>().unwrap())
                    }
                }
            }
        }
        let combined_number =
            format!("{}{}", first_number.unwrap(), last_number.unwrap()).parse::<i32>();
        total += combined_number.unwrap();
    }

    println!("Total: {}", total);
}
